### DISCUSSION TITLE
Modify the loss function

### DISCUSSION THREAD

**Discussion by mmatabuena** [kogalur/randomForestSRC#446] [General]
Hello everyone,

First of all, congratulations on the outstanding work — this is an excellent package. We are particularly interested in exploring new extensions of the algorithm motivated by various applications.

We have two specific questions:

1 How can we incorporate different weights for each observation within the code?
2 Is it possible to modify the objective function to use alternative distances, such as the energy distance?

Thank you very much for your attention, and we look forward to your guidance.

Best regards,

**Comment by ishwaran** [comment]
1. One way to achieve this would be using the option `case.wt` which can be used to over-sample a case in the bootstrap.
2. Regarding modifying the "objective function", the best way to achieve this probably would be to create your own custom splitting rule.  This is because random forests doesn't directly minimize an objective function (like gradient boosting for example) but rather reduces empirical risk by reducing tree impurity.

- Regarding 2, here's what the help file says about this:

Custom splitting: All families except unsupervised are available for user defined custom splitting. Some basic C-programming skills are required. The harness for defining these rules is in splitCustom.c. In this file we give examples of how to code rules for regression, classification, survival, and competing risk. Each family can support up to sixteen custom split rules. Specifying splitrule="custom" or splitrule="custom1" will trigger the first split rule for the family defined by the training data set. Multivariate families will need a custom split rule for both regression and classification. In the examples, we demonstrate how the user is presented with the node specific membership. The task is then to define a split statistic based on that membership. Take note of the instructions in splitCustom.c on how to register the custom split rules. It is suggested that the existing custom split rules be kept in place for reference and that the user proceed to develop splitrule="custom2" and so on. The package must be recompiled and installed for the custom split rules to become available.


- Here's a discussion that could be useful https://github.com/kogalur/randomForestSRC/discussions/197

**Reply by yatingzou-rgb** [reply]
Thanks to your response! Related to the weighing question:

How would you suggest modifying randomForestSRC.c to customize construction of the subsamples for bagged trees? 

For example, if I'd like to utilize weights in the construction, would it be easier to use a new c++ script to replace parts of the existing functionality? Any parts of randomForestSRC.c I should particularly focus on?

Best regards
